### PR TITLE
Support custom options for LLVM build

### PR DESCRIPTION
**Pull request by tmandry**
The LLVM build has a lot of options that rustbuild doesn't need to know about. We should allow the user to customize the LLVM build directly.

Here are some [example customizations][recipe] we'd like to do.

[recipe]: https://fuchsia.googlesource.com/infra/recipes/+/90105e5e4e37b0441c8dde538df54a55f79b3d22/recipes/contrib/clang_toolchain.py#579

### CODE REVIEW

File: config.toml.example [hunk 1]
```diff
@@ -157,6 +157,9 @@ changelog-seen = 2
 # Whether to build the clang compiler.
 #clang = false
 
+# Custom config flags to pass to the LLVM build.
```
Comment (Mark-Simulacrum) [inline config.toml.example:160]:
```suggestion
# Custom CMAKE defines to set when building LLVM.
```

**Comment by Mark-Simulacrum** [comment]
Hm. So, part of me is enthusiastic about supporting something like this, on the other hand, I wonder if it's worth some kind of disclaimer of "at your own risk" or similar -- part of the strategy I've at least sort of tried to steer by is that if folks are trying to extra-customize LLVM, they may just want to build out of tree entirely and provide an llvm-config. (Obviously, that has its own problems).

But I think this is sufficiently general and minimally invasive that I am in general on board with it -- I do wonder if you might be able to ask around and/or have some ideas on other projects embedding LLVM to see if there's some general guidance around the appropriate way to expose this? In particular, with my relatively limited knowledge I'm wondering if just CMAKE defines are typically enough? Or do we need support for general environment configuration / something beyond that?

I will also add that if we do land this it's pretty likely we'll want to drop support for at least some of the existing flag in config.toml that we aren't ourselves interested in using/supporting (e.g., polly build).

**Comment by Mark-Simulacrum** [comment]
cc https://github.com/rust-lang/rust/pull/93629 and https://github.com/rust-lang/rust/pull/93640 as well, which seem like prime PRs I would "not accept" in favor of using this feature, though as noted on one of those I might have pushed for just leaving LLVM builds external as well.

@catap -- curious if you feel this PR would be enough for you or it's missing something you'd need?

**Comment by catap** [comment]
@Mark-Simulacrum definitely!

File: src/bootstrap/config.rs [hunk 2]
```diff
@@ -807,6 +809,7 @@ impl Config {
             config.llvm_allow_old_toolchain = llvm.allow_old_toolchain.unwrap_or(false);
             config.llvm_polly = llvm.polly.unwrap_or(false);
             config.llvm_clang = llvm.clang.unwrap_or(false);
+            config.llvm_build_config = llvm.build_config.unwrap_or(Default::default());
```
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:812]: This probably wants a gate below in the llvm_from_ci! list as well.

**Comment by Mark-Simulacrum** [comment]
r=me with the comment adjusted and llvm_from_ci! gate added.

**Comment by jonhoo** [comment]
I _think_ this also needs an update to `src/bootstrap/configure.py` to allow setting this property through an argument, since as far as I can tell it doesn't have a meaningful way to set dictionaries at the moment.

**Comment by jonhoo** [comment]
Ah, or perhaps it'll just work by including the key for the dictionary in `--set` (so `--set llvm.build-config.LLVM_ENABLE_ZLIB=FORCE_ON` for example)?

**Comment by catap** [comment]
@jonhoo you may use `./configure --set='llvm.build-config={A = 2, C = D}'`

**Comment by Mark-Simulacrum** [comment]
I would expect --set llvm.config.foo=bar to work - do we have something hard coded to two elements in the path?

**Comment by catap** [comment]
@Mark-Simulacrum unfortently it is a bit more tricky. Special when someone needs a few LLVM options. Anyway, some ugly configure set make the job done.

**Comment by jonhoo** [comment]
Yeah, I just tried:
```
--set "llvm.build-config.LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1" \
--set "llvm.build-config.LLVM_ENABLE_ZLIB=FORCE_ON"
```
and got
```
Traceback (most recent call last):
  File "./src/bootstrap/configure.py", line 467, in <module>
    configure_section(sections[section_key], section_config)
  File "./src/bootstrap/configure.py", line 447, in configure_section
    lines[i] = "{} = {}".format(key, to_toml(value))
  File "./src/bootstrap/configure.py", line 436, in to_toml
    raise RuntimeError('no toml')
RuntimeError: no toml
```

**Comment by catap** [comment]
@jonhoo `--set='llvm.build-config={LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1,LLVM_ENABLE_ZLIB=FORCE_ON}'` works as you expected

**Comment by jonhoo** [comment]
I can also confirm that I get the same error even if I just set a single key this way. Will try your way now @catap.

**Comment by jonhoo** [comment]
@catap That doesn't look quite right either. It gives me
```toml
# Custom CMake defines to set when building LLVM.
build-config = '{LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1,LLVM_ENABLE_ZLIB=FORCE_ON}'
```
Which sets `build-config` to a _string_, not a dictionary, which will then fail to deserialize into a `Map` when the bootstrap logic picks it up.

**Comment by catap** [comment]
@jonhoo my command depends on shell which you're using :)

**Comment by jonhoo** [comment]
I don't think that's true — I'm running it in bash, and I think that's the right quoting for the command. The problem is that `set` expects the value to always be set to either a string, a list, or bool. There's no support for emitting it as a dict.

**Comment by catap** [comment]
@jonhoo let me dig, how I've used it. You're right.

**Comment by jonhoo** [comment]
Separately, I also noticed that the way this is currently implemented, `llvm.build-config` _doesn't_ apply to the LLD and sanitizer buildes. I wonder if

https://github.com/rust-lang/rust/blob/89adcc636f94d34a6fc90fa117e28ddf6be7b983/src/bootstrap/native.rs#L374-L376

should move to

https://github.com/rust-lang/rust/blob/89adcc636f94d34a6fc90fa117e28ddf6be7b983/src/bootstrap/native.rs#L415

which all of the three LLVM-based steps use?

**Pull request merged**